### ISSUE TITLE
HostRegexp erroneously rejects hosts with capital letters

### ISSUE THREAD

**Issue by bgandon** [containous/traefik#3930]
### Do you want to request a *feature* or report a *bug*?

This is a bug report for both `traefik` and the `mux` library.

### What did you do?

When using Træfik in front of Cloud foundry, I use a catch-all frontend rule, using a `dummy` variable:
```
            [frontends.cf-http-frontend.routes.catch-all-http]
              # Workaround for a catch-all rule:
              rule = "HostRegexp: {dummy:[-.0-9A-Z_a-z]+}"
```

When performing a request on the CF smoke-tests app (a basic Ruby web app), it works when the host is lowercased:
```
# curl -k https://smoke-tests.apps.easyfoundry.prototyp.it
Healthy
...
```

But when the host is capitalized, then it fails:
```
# curl -k http://SMOKE-TESTS.apps.easyfoundry.prototyp.it
404 page not found
```

And Træfik logs say `backend not found`:
```
"GET / HTTP/1.1" 404 19 "-" "curl/7.35.0" 10 "backend not found" "/" 0ms
```

### What did you expect to see?

I expect the request with capital letters to succeed, as it does when querying CF without Træfik in front of it:

```
# curl -k http://SMOKE-TESTS.apps.easyfoundry.prototyp.it
Healthy
...
```

### What did you see instead?

When the host is capitalized, the request fails:
```
# curl -k http://SMOKE-TESTS.apps.easyfoundry.prototyp.it
404 page not found
```

And Træfik logs say `backend not found`:
```
"GET / HTTP/1.1" 404 19 "-" "curl/7.35.0" 10 "backend not found" "/" 0ms
```

### Output of `traefik version`: (_What version of Traefik are you using?_)

I use version 1.6.5, as deployed by the [traefik-boshrelease](https://github.com/gstackio/traefik-boshrelease), but the incriminated piece of code hasn't changed in master since then, see below.

```
1.6.5
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

The bosh release deploys and configures Træfik v1.6.5 on Ubuntu, and currently I'm deploying on Ubuntu 14.04, but this is not relevant here.

### If applicable, please paste the log output in DEBUG level (`--logLevel=DEBUG` switch)

Not applicable.

### Code Analysis

The issue is a mix of not lowercasing the `getHost(req)` value in `regexp.go` at [line 163](https://github.com/containous/mux/blob/master/regexp.go#L163), but lowercasing any regexp patterns in `rules.go` at [line 59](https://github.com/containous/traefik/blob/master/rules/rules.go#L59), whereas lowercasing a pattern like `[a-zA-Z]` changes its meaning.

### Workaround

A possible workaround is to use a character classes in the pattern that match uppercase characters, but are immune to lowercasing. For example, the `[-.[:alnum:]_]+` regexp works fine.

So the catch-all route configuration becomes:
```toml
          [frontends.cf-http-frontend.routes]
            [frontends.cf-http-frontend.routes.catch-all-http]
              rule = "HostRegexp: {dummy:[-.[:alnum:]_]+}"
```

I'll try to submit PRs in `traefik` and `mux`, referring to this issue.

**Comment by bgandon** [comment]
Also related to containous/mux#6 PR (now merged), that supersedes my containous/mux#5 PR.

**Comment by juliens** [comment]
Closed by #3931
